File: code_puppy/staged_changes.py
"""Staged changes sandbox for safe edit application.

This module provides a staging area where AI-generated edits are accumulated,
reviewed, diffed, and then explicitly applied or rejected, instead of
immediately mutating project files.

Features:
- Intercept file modifications (create, replace, delete)
- Store pending changes in memory with full context
- Generate combined diffs for review
- Apply or reject changes as a batch
- Integration with existing file modification tools
"""

import difflib
import hashlib
import json
import logging
import os
import tempfile
import threading
import time
import uuid
from collections import OrderedDict
from dataclasses import dataclass, field
from enum import Enum, auto
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

STAGE_DIR = Path(tempfile.gettempdir()) / "code_puppy_staged"


class ChangeType(Enum):
    """Types of staged changes."""

    CREATE = auto()
    REPLACE = auto()
    DELETE_SNIPPET = auto()
    DELETE_FILE = auto()


@dataclass
class StagedChange:
    """A single staged change entry."""

    change_id: str
    change_type: ChangeType
    file_path: str

    # For CREATE: new content
    # For REPLACE: old_str and new_str
    # For DELETE_SNIPPET: snippet to delete
    content: str | None = None
    old_str: str | None = None
    new_str: str | None = None
    snippet: str | None = None

    # Metadata
    created_at: float = field(default_factory=time.time)
    description: str = ""
    applied: bool = False
    rejected: bool = False

    def __post_init__(self):
        """Validate that applied and rejected are not both True."""
        if self.applied and self.rejected:
            raise ValueError("StagedChange cannot be both applied and rejected")

    def to_dict(self) -> dict[str, Any]:
        """Serialize change to dictionary."""
        return {
            "change_id": self.change_id,
            "change_type": self.change_type.name,
            "file_path": self.file_path,
            "content": self.content,
            "old_str": self.old_str,
            "new_str": self.new_str,
            "snippet": self.snippet,
            "created_at": self.created_at,
            "description": self.description,
            "applied": self.applied,
            "rejected": self.rejected,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "StagedChange":
        """Create change from dictionary."""
        return cls(
            change_id=data["change_id"],
            change_type=ChangeType[data["change_type"]],
            file_path=data["file_path"],
            content=data.get("content"),
            old_str=data.get("old_str"),
            new_str=data.get("new_str"),
            snippet=data.get("snippet"),
            created_at=data.get("created_at", time.time()),
            description=data.get("description", ""),
            applied=data.get("applied", False),
            rejected=data.get("rejected", False),
        )


class StagedChangesSandbox:
    """Sandbox for accumulating and managing staged file changes."""

    def __init__(self):
        self._lock = threading.Lock()
        self._changes: OrderedDict[str, StagedChange] = OrderedDict()
        self._enabled: bool = False
        self._session_id: str = self._generate_session_id()
        self._ensure_stage_dir()

    def _generate_session_id(self) -> str:
        """Generate unique session ID."""
        return hashlib.sha256(str(time.time()).encode()).hexdigest()[:16]

    def _ensure_stage_dir(self) -> None:
        """Ensure staging directory exists."""
        STAGE_DIR.mkdir(parents=True, exist_ok=True)

    @property
    def enabled(self) -> bool:
        """Check if staging mode is enabled."""
        return self._enabled

    def enable(self) -> None:
        """Enable staging mode."""
        self._enabled = True
        logger.info("Staged changes mode enabled")

    def disable(self) -> None:
        """Disable staging mode (does not clear staged changes)."""
        self._enabled = False
        logger.info("Staged changes mode disabled")

    def toggle(self) -> bool:
        """Toggle staging mode on/off."""
        self._enabled = not self._enabled
        logger.info(f"Staged changes mode {'enabled' if self._enabled else 'disabled'}")
        return self._enabled

    def add_create(
        self, file_path: str, content: str, description: str = ""
    ) -> StagedChange:
        """Stage a file creation."""
        change = StagedChange(
            change_id=self._generate_change_id(),
            change_type=ChangeType.CREATE,
            file_path=os.path.abspath(file_path),
            content=content,
            description=description or f"Create {os.path.basename(file_path)}",
        )
        with self._lock:
            self._changes[change.change_id] = change
        logger.debug(f"Staged create: {file_path}")
        return change

    def add_replace(
        self, file_path: str, old_str: str, new_str: str, description: str = ""
    ) -> StagedChange:
        """Stage a text replacement."""
        change = StagedChange(
            change_id=self._generate_change_id(),
            change_type=ChangeType.REPLACE,
            file_path=os.path.abspath(file_path),
            old_str=old_str,
            new_str=new_str,
            description=description or f"Replace in {os.path.basename(file_path)}",
        )
        with self._lock:
            self._changes[change.change_id] = change
        logger.debug(f"Staged replace: {file_path}")
        return change

    def add_delete_snippet(
        self, file_path: str, snippet: str, description: str = ""
    ) -> StagedChange:
        """Stage a snippet deletion."""
        change = StagedChange(
            change_id=self._generate_change_id(),
            change_type=ChangeType.DELETE_SNIPPET,
            file_path=os.path.abspath(file_path),
            snippet=snippet,
            description=description or f"Delete from {os.path.basename(file_path)}",
        )
        with self._lock:
            self._changes[change.change_id] = change
        logger.debug(f"Staged delete snippet: {file_path}")
        return change

    def _generate_change_id(self) -> str:
        """Generate unique change ID using UUID4."""
        return uuid.uuid4().hex  # 32 hex chars, guaranteed unique

    def get_staged_changes(self, include_applied: bool = False) -> list[StagedChange]:
        """Get all pending staged changes."""
        with self._lock:
            if include_applied:
                return list(self._changes.values())
            return [
                c for c in self._changes.values() if not c.applied and not c.rejected
            ]

    def get_changes_for_file(self, file_path: str) -> list[StagedChange]:
        """Get all staged changes for a specific file."""
        abs_path = os.path.abspath(file_path)
        with self._lock:
            return [
                c
                for c in self._changes.values()
                if c.file_path == abs_path and not c.applied and not c.rejected
            ]

    def clear(self) -> None:
        """Clear all staged changes."""
        with self._lock:
            count = len(self._changes)
            self._changes.clear()
        logger.info(f"Cleared {count} staged changes")

    def remove_change(self, change_id: str) -> bool:
        """Remove a specific change by ID."""
        with self._lock:
            if change_id in self._changes:
                del self._changes[change_id]
                logger.debug(f"Removed change {change_id}")
                return True
        return False

    def count(self) -> int:
        """Count pending staged changes."""
        return len(self.get_staged_changes())

    def is_empty(self) -> bool:
        """Check if no pending staged changes."""
        return self.count() == 0

    def generate_diff(
        self, change: StagedChange, file_cache: dict[str, str] | None = None
    ) -> str:
        """Generate diff for a single change."""
        if change.change_type == ChangeType.CREATE:
            return self._diff_for_create(change)
        elif change.change_type == ChangeType.REPLACE:
            return self._diff_for_replace(change, file_cache)
        elif change.change_type == ChangeType.DELETE_SNIPPET:
            return self._diff_for_delete_snippet(change, file_cache)
        return ""

    def _diff_for_create(self, change: StagedChange) -> str:
        """Generate diff for file creation."""
        content = change.content or ""
        lines = content.splitlines(keepends=True)
        diff = difflib.unified_diff(
            [],
            lines,
            fromfile="/dev/null",
            tofile=f"b/{os.path.basename(change.file_path)}",
            lineterm="",
        )
        return "".join(diff)

    def _diff_for_replace(
        self, change: StagedChange, file_cache: dict[str, str] | None = None
    ) -> str:
        """Generate diff for text replacement."""
        old_str = change.old_str or ""
        new_str = change.new_str or ""

        # Check cache first, then read current file content if it exists
        if file_cache is not None and change.file_path in file_cache:
            original_content = file_cache[change.file_path]
        elif os.path.exists(change.file_path):
            with open(
                change.file_path, "r", encoding="utf-8", errors="surrogateescape"
            ) as f:
                original_content = f.read()
                if file_cache is not None:
                    file_cache[change.file_path] = original_content
        else:
            original_content = ""

        # Apply the replacement
        if old_str in original_content:
            new_content = original_content.replace(old_str, new_str, 1)
        else:
            new_content = original_content

        original_lines = original_content.splitlines(keepends=True)
        new_lines = new_content.splitlines(keepends=True)

        diff = difflib.unified_diff(
            original_lines,
            new_lines,
            fromfile=f"a/{os.path.basename(change.file_path)}",
            tofile=f"b/{os.path.basename(change.file_path)}",
            lineterm="",
        )
        return "".join(diff)

    def _diff_for_delete_snippet(
        self, change: StagedChange, file_cache: dict[str, str] | None = None
    ) -> str:
        """Generate diff for snippet deletion."""
        snippet = change.snippet or ""

        # Check cache first, then read current file content if it exists
        if file_cache is not None and change.file_path in file_cache:
            original_content = file_cache[change.file_path]
        elif os.path.exists(change.file_path):
            with open(
                change.file_path, "r", encoding="utf-8", errors="surrogateescape"
            ) as f:
                original_content = f.read()
                if file_cache is not None:
                    file_cache[change.file_path] = original_content
        else:
            original_content = ""

        # Apply the deletion
        if snippet in original_content:
            new_content = original_content.replace(snippet, "", 1)
        else:
            new_content = original_content

        original_lines = original_content.splitlines(keepends=True)
        new_lines = new_content.splitlines(keepends=True)

        diff = difflib.unified_diff(
            original_lines,
            new_lines,
            fromfile=f"a/{os.path.basename(change.file_path)}",
            tofile=f"b/{os.path.basename(change.file_path)}",
            lineterm="",
        )
        return "".join(diff)

    def generate_combined_diff(self) -> str:
        """Generate combined diff for all pending changes."""
        changes = self.get_staged_changes()
        if not changes:
            return ""

        # Cache file contents to avoid repeated I/O
        file_cache: dict[str, str] = {}

        diffs = []
        for change in changes:
            diff = self.generate_diff(change, file_cache)
            if diff:
                diffs.append(f"# {change.description} ({change.change_id})\n{diff}")

        return "\n\n".join(diffs)

    def preview_changes(self) -> dict[str, str]:
        """Get preview of all changes grouped by file."""
        changes = self.get_staged_changes()
        file_changes: dict[str, list[StagedChange]] = {}

        for change in changes:
            if change.file_path not in file_changes:
                file_changes[change.file_path] = []
            file_changes[change.file_path].append(change)

        result = {}
        for file_path, changes in file_changes.items():
            # Cache file contents per file group to avoid repeated I/O
            file_cache: dict[str, str] = {}
            diffs = [self.generate_diff(c, file_cache) for c in changes]
            result[file_path] = "\n\n".join(diffs)

        return result

    def save_to_disk(self) -> Path:
        """Save staged changes to disk for persistence.

        Uses atomic write-to-tmp-then-rename to avoid corruption if interrupted.

        # FIXME(code-puppy-ctj.5): Triple-write bug — data block repeated 3x.
        # The Elixir port fixes this by writing once.
        """
        self._ensure_stage_dir()
        save_path = STAGE_DIR / f"{self._session_id}.json"

        tmp_path = STAGE_DIR / f"{self._session_id}.json.tmp"

        with self._lock:
            data = {
                "session_id": self._session_id,
                "enabled": self._enabled,
                "changes": [c.to_dict() for c in self._changes.values()],
                "saved_at": time.time(),
            }

        # Atomic write: write to temp file first, then rename
        with open(tmp_path, "w") as f:
            json.dump(data, f, indent=2)
        os.replace(tmp_path, save_path)

        logger.info(f"Saved staged changes to {save_path}")
        return save_path

    def load_from_disk(self, session_id: str | None = None) -> bool:
        """Load staged changes from disk."""
        load_id = session_id or self._session_id
        load_path = STAGE_DIR / f"{load_id}.json"

        if not load_path.exists():
            return False

        try:
            with open(load_path, "r") as f:
                data = json.load(f)

            with self._lock:
                self._session_id = data.get("session_id", load_id)
                self._enabled = data.get("enabled", False)
                loaded_changes = [
                    StagedChange.from_dict(c) for c in data.get("changes", [])
                ]
                self._changes = OrderedDict((c.change_id, c) for c in loaded_changes)

            logger.info(f"Loaded {len(self._changes)} staged changes from {load_path}")
            return True
        except Exception as e:
            logger.error(f"Failed to load staged changes: {e}")
            return False

    def get_summary(self) -> dict[str, Any]:
        """Get summary of staged changes."""
        changes = self.get_staged_changes()

        by_type: dict[str, int] = {}
        by_file: dict[str, int] = {}

        for change in changes:
            type_name = change.change_type.name
            by_type[type_name] = by_type.get(type_name, 0) + 1
            by_file[change.file_path] = by_file.get(change.file_path, 0) + 1

        return {
            "total": len(changes),
            "by_type": by_type,
            "by_file": len(by_file),
            "files": list(by_file.keys()),
            "enabled": self._enabled,
            "session_id": self._session_id,
        }


# Global sandbox instance
_sandbox: StagedChangesSandbox | None = None


def get_sandbox() -> StagedChangesSandbox:
    """Get the global staged changes sandbox."""
    global _sandbox
    if _sandbox is None:
        _sandbox = StagedChangesSandbox()
    return _sandbox


def reset_sandbox() -> StagedChangesSandbox:
    """Reset and create new sandbox."""
    global _sandbox
    _sandbox = StagedChangesSandbox()
    return _sandbox


# Convenience functions for external use


def is_staging_enabled() -> bool:
    """Check if staging mode is enabled."""
    return get_sandbox().enabled


def stage_create(file_path: str, content: str, description: str = "") -> StagedChange:
    """Stage a file creation."""
    return get_sandbox().add_create(file_path, content, description)


def stage_replace(
    file_path: str, old_str: str, new_str: str, description: str = ""
) -> StagedChange:
    """Stage a text replacement."""
    return get_sandbox().add_replace(file_path, old_str, new_str, description)


def stage_delete_snippet(
    file_path: str, snippet: str, description: str = ""
) -> StagedChange:
    """Stage a snippet deletion."""
    return get_sandbox().add_delete_snippet(file_path, snippet, description)


def get_staged_count() -> int:
    """Get count of pending staged changes."""
    return get_sandbox().count()


def clear_staged() -> None:
    """Clear all staged changes."""
    get_sandbox().clear()

File: elixir/code_puppy_control/lib/code_puppy_control/tools/staged_changes.ex
defmodule CodePuppyControl.Tools.StagedChanges do
  @moduledoc """
  Staged changes sandbox — diff-preview gatekeeper system.

  Ports code_puppy/staged_changes.py to Elixir with full parity:
  intercept, store, diff, preview, apply/reject, persist/restore.

  GenServer manages ETS table `:staged_changes` for concurrent reads.

  ## Security

  - All staging operations validate paths via `FileOps.Security.validate_path/2`
    to block sensitive paths (SSH keys, cloud credentials, /etc, etc.)
  - Apply operations route through `FileModifications.SafeWrite` and
    `FileModifications.FileLock` for symlink-safe atomic writes and
    per-file concurrency serialization
  - Staged tools are **slash-only** — not exposed to agents (see `Tools` module)

  ## Tool Exposure Decision (code-puppy-ctj.5)

  Staged change tool modules are intentionally **NOT** registered in the
  Tool.Registry default_modules list. They are slash-only — accessible only
  through the `/staged` command, not as agent-callable tools. This is a
  deliberate design choice: staging is a **user-review mechanism**, and
  allowing agents to invoke staged tools directly would bypass the human
  review intent. The `register_all/0` function exists for testing only.
  """

  use GenServer
  require Logger
  alias CodePuppyControl.Tool.Registry
  alias CodePuppyControl.Tools.StagedChanges.StagedChange
  alias CodePuppyControl.Tools.StagedChanges.{Applier, Diff}
  alias CodePuppyControl.FileOps.Security

  @table :staged_changes
  @stage_dir Path.join(System.tmp_dir!(), "code_puppy_staged")

  # ── Client API ──────────────────────────────────────────────────────────

  def start_link(opts \\ []),
    do: GenServer.start_link(__MODULE__, opts, name: __MODULE__)

  def enabled?, do: GenServer.call(__MODULE__, :enabled?)
  def enable, do: GenServer.call(__MODULE__, :enable)
  def disable, do: GenServer.call(__MODULE__, :disable)

  @doc """
  Toggle staging mode on/off. Returns the new enabled state.
  Matches Python `toggle()`.
  """
  def toggle, do: GenServer.call(__MODULE__, :toggle)

  @doc """
  Stage a file creation.

  Validates the file path against sensitive path rules.
  Returns `{:ok, change}` or `{:error, reason}`.
  """
  def add_create(fp, content, desc \\ ""),
    do: GenServer.call(__MODULE__, {:add_create, fp, content, desc})

  @doc """
  Stage a text replacement.

  Validates the file path against sensitive path rules.
  Returns `{:ok, change}` or `{:error, reason}`.
  """
  def add_replace(fp, old, new, desc \\ ""),
    do: GenServer.call(__MODULE__, {:add_replace, fp, old, new, desc})

  @doc """
  Stage a snippet deletion.

  Validates the file path against sensitive path rules.
  Returns `{:ok, change}` or `{:error, reason}`.
  """
  def add_delete_snippet(fp, snip, desc \\ ""),
    do: GenServer.call(__MODULE__, {:add_delete_snippet, fp, snip, desc})

  @doc """
  Stage a file deletion (DELETE_FILE change type).

  Validates the file path against sensitive path rules.
  Returns `{:ok, change}` or `{:error, reason}`.
  """
  def add_delete_file(fp, desc \\ ""),
    do: GenServer.call(__MODULE__, {:add_delete_file, fp, desc})

  @doc "Get pending staged changes (applied/rejected excluded by default).
  Pass `include_applied: true` to return all."
  def get_staged_changes(opts \\ []) do
    include_applied = Keyword.get(opts, :include_applied, false)

    case :ets.whereis(@table) do
      :undefined ->
        []

      _ ->
        changes =
          @table
          |> :ets.tab2list()
          |> Enum.map(fn {_, c} -> c end)

        if include_applied do
          changes |> Enum.sort_by(& &1.created_at)
        else
          changes
          |> Enum.reject(&(&1.applied or &1.rejected))
          |> Enum.sort_by(& &1.created_at)
        end
    end
  end

  @doc """
  Get staged changes for a specific file (pending only).
  Matches Python `get_changes_for_file`.
  """
  def get_changes_for_file(file_path) do
    abs_path = Path.expand(file_path)

    get_staged_changes()
    |> Enum.filter(&(&1.file_path == abs_path))
  end

  @doc """
  Count of pending staged changes.
  """
  def count, do: get_staged_changes() |> length()

  @doc """
  Check if no pending staged changes.
  Matches Python `is_empty`.
  """
  def is_empty?, do: count() == 0

  @doc """
  Remove a specific change by ID.

  Returns `true` if the change was found and removed, `false` otherwise.
  This provides boolean parity with the expected remove-change semantics.
  """
  def remove_change(id), do: GenServer.call(__MODULE__, {:remove_change, id})

  @doc """
  Clear all staged changes.
  """
  def clear, do: GenServer.call(__MODULE__, :clear)

  @doc """
  Get session ID for this sandbox.
  """
  def session_id, do: GenServer.call(__MODULE__, :session_id)

  @doc """
  Generate combined diff for all pending changes.
  Uses file I/O cache to avoid repeated reads (matches Python `generate_combined_diff`).
  """
  def get_combined_diff do
    changes = get_staged_changes()

    if changes == [] do
      ""
    else
      # Cache file contents to avoid repeated I/O (Python parity)
      {diffs, _final_cache} =
        Enum.reduce(changes, {[], %{}}, fn c, {acc, cache} ->
          {diff, new_cache} = Diff.gen_diff_cached(c, cache)

          if diff == "" do
            {acc, new_cache}
          else
            {[{c.description, c.change_id, diff} | acc], new_cache}
          end
        end)

      diffs
      |> Enum.reverse()
      |> Enum.map(fn {desc, id, d} -> "# #{desc || "change"} (#{id})\n#{d}" end)
      |> Enum.join("\n\n")
    end
  end

  @doc "Preview changes grouped by file."
  def preview_changes do
    changes = get_staged_changes()

    # Group by file_path
    grouped = Enum.group_by(changes, & &1.file_path)

    Enum.map(grouped, fn {file_path, file_changes} ->
      {diffs, _cache} =
        Enum.reduce(file_changes, {[], %{}}, fn c, {acc, cache} ->
          {diff, new_cache} = Diff.gen_diff_cached(c, cache)

          if diff == "" do
            {acc, new_cache}
          else
            {[diff | acc], new_cache}
          end
        end)

      combined = diffs |> Enum.reverse() |> Enum.join("\n\n")
      {file_path, combined}
    end)
    |> Map.new()
  end

  @doc "Get summary of staged changes (by_type, by_file, etc.)."
  def get_summary do
    changes = get_staged_changes()

    by_type =
      Enum.reduce(changes, %{}, fn c, acc ->
        key = Atom.to_string(c.change_type)
        Map.update(acc, key, 1, &(&1 + 1))
      end)

    by_file =
      Enum.reduce(changes, %{}, fn c, acc ->
        Map.update(acc, c.file_path, 1, &(&1 + 1))
      end)

    %{
      total: length(changes),
      by_type: by_type,
      by_file: map_size(by_file),
      files: Map.keys(by_file),
      enabled: GenServer.call(__MODULE__, :enabled?),
      session_id: GenServer.call(__MODULE__, :session_id)
    }
  end

  @doc "Save staged changes to disk (atomic write). Fixes Python triple-write bug."
  def save_to_disk do
    GenServer.call(__MODULE__, :save_to_disk)
  end

  @doc "Load staged changes from disk. Returns `true` on success, `false` on failure."
  def load_from_disk(session_id \\ nil) do
    GenServer.call(__MODULE__, {:load_from_disk, session_id})
  end

  def apply_all, do: GenServer.call(__MODULE__, :apply_all, 30_000)
  def reject_all, do: GenServer.call(__MODULE__, :reject_all)

  @doc """
  Register all staged changes tool modules with the Tool Registry.

  **NOTE:** This function is for testing only. Staged change tools are
  **slash-only** — not exposed to agents. See module doc for rationale.
  """
  def register_all do
    alias CodePuppyControl.Tools.StagedChanges.Tools
    alias CodePuppyControl.Tool.Registry

    [
      Tools.StageCreateTool,
      Tools.StageReplaceTool,
      Tools.StageDeleteSnippetTool,
      Tools.StageDeleteFileTool,
      Tools.GetStagedDiffTool,
      Tools.ApplyStagedTool,
      Tools.RejectStagedTool
    ]
    |> Enum.reduce({:ok, 0}, fn m, {:ok, acc} ->
      case Registry.register(m) do
        :ok -> {:ok, acc + 1}
        _ -> {:ok, acc}
      end
    end)
  end

  # ── GenServer Callbacks ─────────────────────────────────────────────────

  @impl true
  def init(_opts) do
    table =
      try do
        :ets.new(@table, [
          :named_table,
          :set,
          :public,
          read_concurrency: true,
          write_concurrency: true
        ])
      rescue
        ArgumentError -> @table
      end

    session_id = generate_session_id()
    ensure_stage_dir()

    {:ok, %{table: table, enabled: false, session_id: session_id}}
  end

  @impl true
  def handle_call(:enabled?, _, s), do: {:reply, s.enabled, s}
  @impl true
  def handle_call(:enable, _, s), do: {:reply, :ok, %{s | enabled: true}}
  @impl true
  def handle_call(:disable, _, s), do: {:reply, :ok, %{s | enabled: false}}

  @impl true
  def handle_call(:toggle, _, s) do
    new_enabled = not s.enabled
    {:reply, new_enabled, %{s | enabled: new_enabled}}
  end

  @impl true
  def handle_call(:session_id, _, s), do: {:reply, s.session_id, s}

  @impl true
  def handle_call({:add_create, fp, content, desc}, _, s) do
    with {:ok, _} <- validate_staged_path(fp, "stage_create") do
      c = mk(:create, fp, content: content, description: desc)
      :ets.insert(@table, {c.change_id, c})
      {:reply, {:ok, c}, s}
    else
      {:error, reason} -> {:reply, {:error, reason}, s}
    end
  end

  @impl true
  def handle_call({:add_replace, fp, old, new, desc}, _, s) do
    with {:ok, _} <- validate_staged_path(fp, "stage_replace") do
      c = mk(:replace, fp, old_str: old, new_str: new, description: desc)
      :ets.insert(@table, {c.change_id, c})
      {:reply, {:ok, c}, s}
    else
      {:error, reason} -> {:reply, {:error, reason}, s}
    end
  end

  @impl true
  def handle_call({:add_delete_snippet, fp, snip, desc}, _, s) do
    with {:ok, _} <- validate_staged_path(fp, "stage_delete_snippet") do
      c = mk(:delete_snippet, fp, snippet: snip, description: desc)
      :ets.insert(@table, {c.change_id, c})
      {:reply, {:ok, c}, s}
    else
      {:error, reason} -> {:reply, {:error, reason}, s}
    end
  end

  @impl true
  def handle_call({:add_delete_file, fp, desc}, _, s) do
    with {:ok, _} <- validate_staged_path(fp, "stage_delete_file") do
      c = mk(:delete_file, fp, description: desc)
      :ets.insert(@table, {c.change_id, c})
      {:reply, {:ok, c}, s}
    else
      {:error, reason} -> {:reply, {:error, reason}, s}
    end
  end

  @impl true
  def handle_call({:remove_change, id}, _, s) do
    # Return boolean: true if found and deleted, false if not found
    case :ets.lookup(@table, id) do
      [{^id, _}] ->
        :ets.delete(@table, id)
        {:reply, true, s}

      [] ->
        {:reply, false, s}
    end
  end

  @impl true
  def handle_call(:clear, _, s) do
    :ets.delete_all_objects(@table)
    {:reply, :ok, s}
  end

  @impl true
  def handle_call(:reject_all, _, s) do
    changes = get_staged_changes()
    for c <- changes, do: :ets.insert(@table, {c.change_id, %{c | rejected: true}})
    {:reply, length(changes), s}
  end

  @impl true
  def handle_call(:apply_all, _, s) do
    changes = get_staged_changes()

    if changes == [] do
      {:reply, {:ok, 0}, s}
    else
      result =
        Enum.reduce_while(changes, {:ok, 0}, fn c, {:ok, acc} ->
          case Applier.apply_change(c) do
            :ok ->
              :ets.insert(@table, {c.change_id, %{c | applied: true}})
              {:cont, {:ok, acc + 1}}

            {:error, r} ->
              {:halt, {:error, "Failed #{c.change_id}: #{r}"}}
          end
        end)

      {:reply, result, s}
    end
  end

  @impl true
  def handle_call(:save_to_disk, _, s) do
    ensure_stage_dir()
    save_path = Path.join(@stage_dir, "#{s.session_id}.json")
    tmp_path = Path.join(@stage_dir, "#{s.session_id}.json.tmp")

    data = %{
      "session_id" => s.session_id,
      "enabled" => s.enabled,
      "changes" => get_staged_changes(include_applied: true) |> Enum.map(&StagedChange.to_map/1),
      "saved_at" => System.system_time(:second)
    }

    # Atomic write: write to temp file first, then rename
    # Fixes the Python triple-write bug where data block was repeated 3x
    case File.write(tmp_path, Jason.encode!(data, pretty: true)) do
      :ok ->
        case File.rename(tmp_path, save_path) do
          :ok ->
            Logger.info("Saved staged changes to #{save_path}")
            {:reply, {:ok, save_path}, s}

          {:error, reason} ->
            File.rm(tmp_path)
            {:reply, {:error, "Rename failed: #{reason}"}, s}
        end

      {:error, reason} ->
        {:reply, {:error, "Write failed: #{reason}"}, s}
    end
  end

  @impl true
  def handle_call({:load_from_disk, nil}, _, s) do
    handle_call({:load_from_disk, s.session_id}, nil, s)
  end

  @impl true
  def handle_call({:load_from_disk, session_id}, _, s) do
    load_path = Path.join(@stage_dir, "#{session_id}.json")

    case File.read(load_path) do
      {:ok, raw} ->
        case Jason.decode(raw) do
          {:ok, data} ->
            # Robust deserialization: validate "changes" is a list
            raw_changes = Map.get(data, "changes", [])

            {loaded_changes, skipped} =
              if is_list(raw_changes) do
                Enum.reduce(raw_changes, {[], 0}, fn entry, {acc, skip_count} ->
                  case StagedChange.from_map(entry) do
                    {:ok, c} ->
                      {[c | acc], skip_count}

                    {:error, reason} ->
                      Logger.warning("Skipping malformed staged change: #{reason}")
                      {acc, skip_count + 1}
                  end
                end)
              else
                Logger.warning(
                  "Ignoring malformed 'changes' field (expected list, got #{inspect(raw_changes)}), loading 0 changes"
                )

                {[], 0}
              end

            loaded_changes = Enum.reverse(loaded_changes)

            if skipped > 0 do
              Logger.warning("Skipped #{skipped} malformed staged changes during load")
            end

            # Rebuild ETS from loaded changes
            :ets.delete_all_objects(@table)

            for c <- loaded_changes do
              :ets.insert(@table, {c.change_id, c})
            end

            new_session_id = Map.get(data, "session_id", session_id)
            new_enabled = Map.get(data, "enabled", false)

            Logger.info("Loaded #{length(loaded_changes)} staged changes from #{load_path}")

            {:reply, true, %{s | session_id: new_session_id, enabled: new_enabled}}

          {:error, reason} ->
            Logger.error("Failed to parse staged changes JSON: #{inspect(reason)}")
            {:reply, false, s}
        end

      {:error, _} ->
        {:reply, false, s}
    end
  end

  @impl true
  def handle_info(_, s), do: {:noreply, s}

  # ── Private helpers ──────────────────────────────────────────────────────

  defp generate_session_id do
    :crypto.hash(:sha256, :erlang.term_to_binary(System.system_time(:millisecond)))
    |> Base.encode16(case: :lower)
    |> String.slice(0, 16)
  end

  defp ensure_stage_dir do
    File.mkdir_p!(@stage_dir)
  end

  defp validate_staged_path(fp, operation) do
    Security.validate_path(fp, operation)
  end

  defp mk(type, fp, opts) do
    desc = if (d = opts[:description]) != nil and d != "", do: d, else: default_desc(type, fp)

    StagedChange.new(
      change_id: id(),
      change_type: type,
      file_path: Path.expand(fp),
      content: opts[:content],
      old_str: opts[:old_str],
      new_str: opts[:new_str],
      snippet: opts[:snippet],
      description: desc,
      # Use microsecond timestamp for stable insertion-order sorting
      created_at: System.system_time(:microsecond),
      applied: false,
      rejected: false
    )
  end

  defp default_desc(:create, fp), do: "Create #{Path.basename(fp)}"
  defp default_desc(:replace, fp), do: "Replace in #{Path.basename(fp)}"
  defp default_desc(:delete_snippet, fp), do: "Delete from #{Path.basename(fp)}"
  defp default_desc(:delete_file, fp), do: "Delete #{Path.basename(fp)}"
  defp default_desc(_, fp), do: "Change #{Path.basename(fp)}"

  defp id, do: :crypto.strong_rand_bytes(16) |> Base.encode16(case: :lower)
end
